39 collect patent citations

india-kerle opened the pull request:
# Description

This PR adds a patent citations collection pipeline. The output (i.e. .json where each key is a patent id and each value is a list of patent ids) should be in identical format to OpenAlex so as to calculate the consolidation-distruption (CD) index. 

I've also gotten rid of the temporary AI genomics patents getters that aren't being used anymore. 

Fixes # (issue)

This PR closes #39 

# Instructions for Reviewer

The main script to run that collects forward and backward citations is:

`python dap_aria_mapping/pipeline/data_collection/paents_citations.py run --production=False`

To test the flow:

`pytest dap_aria_mapping/pipeline/data_collection/tests/test_patents_citations.py`

i've ultimately decided to threshold based on years (2007 and 2017) and _not_ citation types because I think if we wanted to threshold based on types, we really need to talk to a lawyer or patents expert. I tried to google around for fleshed out definitions (a bit more context here - https://docs.google.com/spreadsheets/d/1LtfjECVI5pqqwE7oMw1JbwFcUWhUoHgJH_mJ0flw9Fw/edit#gid=1193216467)  and also reached out to edward but he doesn't know too much about citation types. 

# Please pay special attention to ...

- [x] Did I get the definition of forward vs. backward citations right??

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

Commit:
add patents citation flow

Commit:
add tests and update README

Commit:
update typo

india-kerle committed:
Update README.md

india-kerle committed:
Update README.md

Commit:
get rid of unnessesary import

Commit:
remove temporary oa ai genomics getters

Commit:
citations collection take two

Commit:
add updated readme

Commit:
updated script to only collect for early and late years

Commit:
import pandas outside of steps

Commit:
deal with merge conflict

Commit:
update readme

Commit:
update readme

Commit:
remove unused parameter

Commit:
fixed citations test

Commit:
more efficient query, thresholding, update tests

Commit:
update getters

Commit:
optmise query + query chunk

Commit:
update readme

georgerichardson commented on dap_aria_mapping/pipeline/data_collection/patents_citations.py:
```diff
@@ -0,0 +1,268 @@
+    Returns:
+        str: Base query string.
+    """
+    return "SELECT p.publication_number as focal_id, c.publication_number as patent_citation from `patents-public-data.patents.publications` as p, unnest(citation) c WHERE p.publication_number IN ({})"
```
Minor, but might be more readable to break up long SQL query strings into multiline strings with `"""` or `()`

georgerichardson commented on dap_aria_mapping/pipeline/data_collection/patents_citations.py:
```diff
@@ -0,0 +1,268 @@
+    """
+    ids = publication_numbers if production else publication_numbers[:10]
+
+    publication_numbers_chunks = chunk(ids, chunk_size)
```
see `toolz.partition_all` to never have to write a chunks function again

georgerichardson commented on dap_aria_mapping/pipeline/data_collection/patents_citations.py:
```diff
@@ -0,0 +1,268 @@
+
+        # get list of publication numbers i.e. focal ids from patents data for early and late years
+        logger.info(
+            f"Fetching list of focal ids for year {EARLY_YEAR} and {LATE_YEAR}."
```
In the publication citations flow, I only specified a single year which the user has to vary depending on which year they want to collect. On the one hand your way makes more sense from a reproduction view, but I think it's a little confusing to have a year range that only collects for those specific years. I'd go with a single year and run it twice or an alternative could be to provide the param as a list of years.

georgerichardson commented on dap_aria_mapping/pipeline/data_collection/patents_citations.py:
```diff
@@ -0,0 +1,268 @@
+            upload_obj(
+                self.forward_citations,
+                BUCKET_NAME,
+                "inputs/data_collection/patents/patents_forward_citations.json",
```
Maybe save different years as separate files and include the year in the file name

georgerichardson commented on dap_aria_mapping/pipeline/data_collection/README.md:
```diff
@@ -46,6 +48,26 @@ Output data from these pipelines is stored on S3 at `inputs/data_collection/open
+
+This flow outputs three files stored on S3 at `inputs/data_collection/patents/`:
+
+1. **`patents_backward_citations.json`**: A json where the key is a focal patent id and the value is a list of patent ids that the focal id cites.
+
+2. **`patents_cited_by_citations.json`**: A json where the key is a focal patent id and the value is a list of patent ids that the focal patent id is cited by.
+
+3. **`patents_forward_citations.json`**: A json where the key is a cited by patent id and the value is a list of patents that it cites. A focal id should be present in each dictionary value. 
```
great!

georgerichardson replied:
```suggestion
1. **`patents_backward_citations.json`**: A json where the key is a focal patent id and the value is a list of patent ids that the focal id cites.

2. **`patents_cited_by_citations.json`**: A json where the key is a focal patent id and the value is a list of patent ids that the focal patent id is cited by.

3. **`patents_forward_citations.json`**: A json where the key is a cited by patent id and the value is a list of patents that it cites. At least one focal id will be present in each dictionary value. 
```

I hope we can make this edit with certainty?

georgerichardson commented on dap_aria_mapping/utils/patents.py:
```diff
@@ -0,0 +1,30 @@
+    Returns:
+        List[List[str]]: List of lists of chunked strings.
+    """
+    return [lst[i::n] for i in range(n)]
```
Does chunking by number of chunks make sense here? If the number of results is really large, then you could end up with really big chunks. OTOH you could chunk by number of items per chunk.

georgerichardson reviewed (commented):
Looks like we're collecting the right stuff.

I noted that you `set` the list of cited by IDs. Is any more deduplication needed e.g. by patent family ID for those results or the citations themselves?

india-kerle committed:
Update dap_aria_mapping/pipeline/data_collection/README.md

Co-authored-by: George Richardson <georgerichardson@posteo.net>

Commit:
address george's PR comments

Commit:
address george's PR comments

india-kerle commented:
cool I've addressed your comments @georgerichardson and run the flow in production for the years 2007 and 2017 - let me know if you think I should make final changes

india-kerle committed:
Merge branch 'dev' into 39_patent_citations

india-kerle commented:
do let me know if it's ok to merge @georgerichardson !
